docs/source/examples.daisy_chain_gap.ipynb

```diff
@@ -0,0 +1,746 @@
+    "if outputs.done():\n",
+    "    print(\"Not calculating SOAP, because output file is found\")\n",
+    "    inputs = outputs.to_ConfigSet()\n",
+    "else:\n",
+    "    descriptor_key = \"SOAP\"\n",
+    "    # Descriptor string, just as it would go into quip.\n",
+    "    # dictionary can have a descriptor per species, e.g. \n",
+    "    # descriptor = {\n",
+    "    #   \"H\": \"soap ...\",\n",
+    "    #   \"C\": \"soap ...\"}\n",
+    "    # `None` for dictionary keys just means that the same descriptor is used \n",
+    "    # for all elements. \n",
+    "    descriptor =   {\n",
+    "            None: \"soap l_max=3 n_max=6 cutoff=4 delta=1 covariance_type=dot_product zeta=4 atom_gaussian_width=0.3\"\n",
+    "        }\n",
+    "\n",
+    "    # this function isn't parallelised here, but can be\n",
+    "    # by setting WFL_NUM_PYTHON_SUBPROCESSES or\n",
+    "    # WFL_EXPYRE_INFO\n",
+    "    inputs = wfl.descriptors.quippy.calc(\n",
+    "        inputs=inputs,\n",
+    "        outputs=tmp_outputs,\n",
+    "        descs=descriptor,\n",
+    "        key=descriptor_key,\n",
+    "        per_atom=True\n",
+    "    )\n",
+    "\n",
+    "    # make global average descriptor out of local ones\n",
+    "    for at in inputs:\n",
+    "        at_desc = at.arrays.pop(descriptor_key)\n",
+    "        at_desc = np.sum(at_desc, axis=0)\n",
+    "        at_desc /= np.linalg.norm(at_desc)\n",
+    "        at.info[descriptor_key] = at_desc\n",
+    "        outputs.store(at)\n",
+    "    outputs.close()\n",
+    "    inputs = outputs.to_ConfigSet()"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Sub-select with CUR\n",
+    "\n",
+    "Select diverse structures for training and testing sets with CUR. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/eg475/dev/workflow/wfl/select/by_descriptor.py:197: UserWarning: output is done, returning\n",
+      "  warnings.warn('output is done, returning')\n"
+     ]
+    }
+   ],
+   "source": [
+    "outputs = OutputSpec(\"5.ch.rdkit.md.traj.soap.cur_selection.xyz\")\n",
+    "inputs = wfl.select.by_descriptor.CUR_conf_global(\n",
+    "    inputs=inputs,\n",
+    "    outputs=outputs,\n",
+    "    num=100,                    # target number of environments to pick\n",
+    "    at_descs_info_key=\"SOAP\"\n",
+    ")\n",
+    "\n",
+    "train_fname = \"6.1.train.xyz\"\n",
+    "test_fname = \"6.2.test.xyz\"\n",
+    "gap_fname='gap.xml'\n",
+    "\n",
+    "# process configs before fitting\n",
+    "inputs = list(inputs)\n",
+    "for at in inputs:\n",
+    "\n",
+    "    # set the cell for gap_fit\n",
+    "    at.cell = [50, 50, 50]\n",
+    "\n",
+    "    # gap_fit cannot parse the xyz correctly with the SOAP descriptor\n",
```

Comment:
Do we understand why not?

Reply:
Do you have the file? I could take a look.

Reply:
OK - the immediate problem is that the header line is too long.  That's trivial to fix in `libAtoms`, but it doesn't actually help, because the fortran xyz reader only allows info float fields to be length 1, 3, 6, or 9. I think the new xyz reader fixes that, but I don't know how close we are to using it.  As a result, it chokes on the SOAP field either way.  I think we need to clean up the descriptor in this example, but I'll try to push QUIP to fix this issue.

Reply:
OK, I fixed the random errors, but the test fails on the actual reference quantities. I suggest running it locally with pytest rather than relying on the CI, but I tried to give a more informative error. I'm not sure if the summary outputted by pytest is sufficient, but if you run pytest -s and scroll back you can find what actually failed. In my case it's a quantitative mismatch in the error table, specifically
```
E   AssertionError: Mismatch for ref prop atomization_energy/atom config train measure RMSE ref 0.014974721709894103 != actual 0.01606044820056456
```

Reply:
Clearer reporting is very helpful, thanks. 

It seems that difference is because of different architectures gap_fit has been run on? The gap coefficients and predictions vary  _tiny_ bit from fit to fit, even with a given random seed. Also I get slightly different results when running notebook directly vs with pytest (which I didn't expect?). We could have CI values as reference, but then local tests would fail, so I'll relax the tolerance.

Reply:
It seems that the difference between local notebook and local pytest starts with RDKit geometries, even if they're generated the same whenever the notebook is re-run. I don't think I can set a random seed (Python RDKit is a wrapper to C++). Since this is mostly testing the workflow, I think very loose tolerance (0.1 eV or eV/Å) is the  optimal choice? This should pick out drastic breaking changes...

Reply:
Yeah, sometimes you need loose tolerances, or even testing some other (aggregate or averaged) quantity if you can't control some randomness.  We're having a bad time with that in ACEHAL, for reasons that I understand even less.

If you can reduce generating the initial non-deterministic configuration it to something simple, I can take a look as well.

Reply:
I have tried reading initial structures from file, but MD wasn't deterministic with set random seed. Or at least the notebook and pytest MD trajectories were different, didn't check if they were deterministic from run to run. I'm guessing that there's not enough significant figures to fill a double saved in xyz and the rest is filled with noise? 

Could try saving/loading the initial structures from e.g. ase.db, but not sure that's worth it?

Reply:
The MD isn't deterministic if you're doing any parallelization unless you set the env var `WFL_DETERMINISTIC_HACK` to anything (which the gap rss iter fit test does, but isn't done across the board).  I guess we should consider fixing the non-determinism itself if possible, which I think I know how to do but it'd be tricky.

If that's the only issue, you could try setting that env var.